src/actor/disk.rs

```diff
@@ -0,0 +1,191 @@
+//! An antagonist that exercises disk lifecycle commands (create, delete).
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use oxide_api::types::BlockSize;
+use oxide_api::types::ByteCount;
+use oxide_api::types::DiskCreate;
+use oxide_api::types::DiskSource;
+use oxide_api::types::DiskState;
+use oxide_api::types::Name;
+use oxide_api::ClientDisksExt;
+use tracing::{info, trace};
+
+use crate::util::{fail_if_no_response, sleep_random_ms};
+
+/// The possible actions that this antagonist can take.
+#[derive(Debug, Clone, Copy)]
+enum Action {
+    Wait,
+    Create,
+    Delete,
+}
+
+/// The parameters used to configure a disk antagonist.
+pub struct Params {
+    /// The name of the project to create this antagonist's disk in.
+    pub project: String,
+
+    /// The name of the disk this antagonist should act on.
+    pub disk_name: String,
+}
+
+/// The internal state for a disk antagonist.
+#[derive(Debug)]
+pub(super) struct DiskActor {
+    client: oxide_api::Client,
+    project: String,
+    disk_name: String,
+}
+
+impl DiskActor {
+    /// Creates a new disk antagonist.
+    pub(super) fn new(params: Params) -> Result<Self> {
+        Ok(Self {
+            client: crate::client::get_client(crate::config())?,
+            project: params.project,
+            disk_name: params.disk_name,
+        })
+    }
+
+    /// Gets this actor's disk's current state.
+    ///
+    /// # Return value
+    ///
+    /// - Ok(Some(state)) if the query succeeded.
+    /// - Ok(None) if the query failed with a "not found" error.
+    /// - Err if the query failed for any other reason.
+    async fn get_disk_state(&self) -> Result<Option<DiskState>> {
+        let res = self
+            .client
+            .disk_view()
+            .project(&self.project)
+            .disk(&self.disk_name)
+            .send()
+            .await;
+
+        match res {
+            Ok(response_value) => Ok(Some(response_value.into_inner().state)),
+
+            Err(e) => match &e {
+                oxide_api::Error::InvalidRequest(_)
+                | oxide_api::Error::CommunicationError(_)
+                | oxide_api::Error::InvalidResponsePayload(_)
+                | oxide_api::Error::UnexpectedResponse(_) => Err(e.into()),
+
+                oxide_api::Error::ErrorResponse(response_value) => {
+                    let status = response_value.status();
+
+                    // It's OK if the disk just isn't there. Any other error
+                    // is unexpected.
+                    if status == http::StatusCode::NOT_FOUND {
+                        Ok(None)
+                    } else {
+                        Err(e.into())
+                    }
+                }
+            },
+        }
+    }
+
+    /// Asks to create this actor's disk. The created disk has 5 GB.
```

Comment:
Does this impose any special constraints on the size of the vdevs one has to create if standing up an Omicron dev cluster to run these tests? (I suppose it might not; in the worst case you might just find all your disk creates fail because there's no space.) By comparison, for instance lifecycle operations I explicitly made the instances as small as possible to maximize the number of them I could create concurrently.

Reply:
There's no reason they couldn't go down to the minimum supported size, no. It may even be a good test to run the disks out of space!

Reply:
053dcb2 changes to use Nexus' minimum supported disk size of 1G